multimodal/src/autogluon/multimodal/learners/base.py

```diff
@@ -845,7 +845,8 @@ def get_datamodule_per_run(
         predict_data=None,
         is_train=True,
     ):
-        if self._teacher_learner is not None:
+        if self._teacher_learner is not None and not isinstance(self._teacher_learner, str):
```

Comment:
In what cases teacher_learner be a str when running `get_datamodule_per_run`? In `fit_per_run`, `get_datamodule_per_run` is called after `setup_distillation`, which should have converted teacher_learner into a learner object.

Reply:
```
../../../multimodal/tests/unittests/others_2/test_hpo.py:200:                                                            
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../../multimodal/src/autogluon/multimodal/predictor.py:460: in fit                                                    
    self._learner.fit(                                                                                                   
../../../multimodal/src/autogluon/multimodal/learners/base.py:643: in fit                                                
    fit_returns = self.execute_fit()                                                                                     
../../../multimodal/src/autogluon/multimodal/learners/base.py:549: in execute_fit                                        
    hyperparameter_tune(                                                                                                 
../../../multimodal/src/autogluon/multimodal/utils/hpo.py:229: in hyperparameter_tune                                    
    learner = build_final_learner(                                                                                       
../../../multimodal/src/autogluon/multimodal/utils/hpo.py:125: in build_final_learner                                    
    learner.top_k_average(                                                                                               
../../../multimodal/src/autogluon/multimodal/learners/base.py:1395: in top_k_average                                     
    best_score = self.evaluate(self._tuning_data, metrics=[self._eval_metric_name])[                                     
../../../multimodal/src/autogluon/multimodal/learners/base.py:1806: in evaluate                                          
    outputs = self.predict_per_run(                                                                                      
../../../multimodal/src/autogluon/multimodal/learners/base.py:1722: in predict_per_run                                   
    datamodule = self.get_datamodule_per_run(    
```

Reply:
In predict_per_run the teacher_learner is not set up. And while doing hpo build_final_learner there is no training but only evaluation.

Reply:
Why don't our hpo tests pass fail?

Reply:
A better fix is to move 
    ```
    if self._teacher_learner is not None:
            df_preprocessor = [df_preprocessor, self._teacher_learner._df_preprocessor]
            data_processors = [data_processors, self._teacher_learner._data_processors]
```
inside the `if is_train` because only training needs the teacher_learner.

Reply:
> Why didn't our hpo tests pass fail?

It was using realtime inference.

Reply:
> A better fix is to move
> 
> ```
>     if self._teacher_learner is not None:
>             df_preprocessor = [df_preprocessor, self._teacher_learner._df_preprocessor]
>             data_processors = [data_processors, self._teacher_learner._data_processors]
> ```
> 
> inside the `if is_train` because only training needs the teacher_learner.

Moved the logic inside `is_train`